app/scripts/modules/amazon/src/function/configure/Network.tsx

```diff
@@ -0,0 +1,247 @@
+    return {};
+  }
+
+  private getAvailableSubnets(): IPromise<ISubnet[]> {
+    return SubnetReader.listSubnetsByProvider('aws');
+  }
+
+  private getAvailableSecurityGroups(): IPromise<ISecurityGroupsByAccountSourceData> {
+    return ReactInjector.securityGroupReader.getAllSecurityGroups();
+  }
+  private makeSubnetOptions(availableSubnets: ISubnet[]): ISubnetOption[] {
+    const subOptions: ISubnetOption[] = [];
+    availableSubnets.forEach(s => {
+      const opt: ISubnetOption = {
+        subnetId: s.id,
+        vpcId: s.vpcId,
+      };
+      subOptions.push(opt);
+    });
+    // we have to filter out any duplicate options
+    const uniqueSubOptions = Array.from(new Set(subOptions.map(a => a.subnetId))).map(subnetId => {
+      return subOptions.find(a => a.subnetId === subnetId);
+    });
+    return uniqueSubOptions;
+  }
+
+  public componentDidUpdate() {
+    this.props$.next(this.props);
+  }
+
+  public componentWillUnmount(): void {
+    this.destroy$.next();
+  }
+
+  public componentDidMount(): void {
+    const subnets$ = Promise.resolve(this.getAvailableSubnets())
+      .then((subnets: ISubnet[]) => {
+        subnets.forEach((subnet: ISubnet) => {
+          subnet.label = subnet.id;
+          subnet.deprecated = !!subnet.deprecated;
+          if (subnet.deprecated) {
+            subnet.label += ' (deprecated)';
+          }
+        });
+        return subnets.filter(s => s.label);
+      })
+      .then((subnets: ISubnet[]) => {
+        return this.makeSubnetOptions(subnets);
+      });
+
+    const secGroups$ = Promise.resolve(this.getAvailableSecurityGroups());
+    Observable.combineLatest(subnets$, secGroups$)
+      .takeUntil(this.destroy$)
+      .subscribe(([subnets, securityGroups]) => {
+        return this.setState({ subnets, securityGroups });
+      });
+  }
+
+  private handleSubnetUpdate = (options: Array<Option<string>>) => {
+    const subnetsSelected = options.map(o => o.value);
+    this.props.formik.setFieldValue('subnetIds', subnetsSelected);
+  };
+
+  private handleSecurityGroupsUpdate = (options: Array<Option<string>>) => {
+    const sgSelected = options.map(o => o.value);
+    this.props.formik.setFieldValue('securityGroupIds', sgSelected);
+  };
+
+  private handleVpcChange = (vpcId: string): void => {
+    this.props.formik.setFieldValue('vpcId', vpcId);
+    const { subnets } = this.state;
+    const subs = subnets.filter(function(s: ISubnetOption) {
+      return s.vpcId.includes(vpcId);
+    });
+    this.setState({ subnets: subs });
+  };
+
+  private toSubnetOption = (value: ISubnetOption): Option<string> => {
+    return { value: value.subnetId, label: value.subnetId };
+  };
+
+  private getSecurityGroupsByVpc = (sgs: ISecurityGroupsByAccountSourceData): Array<Option<string>> => {
+    const { values } = this.props.formik;
+    const sgOptions: Array<Option<string>> = [];
+    forOwn(sgs, function(sgByAccount, acc) {
```

Comment:
Would it make sense to just `filter` the `sgs` by all the attributes you listed, then map to `{ value, label }`?

Reply:
I had to use `forOwn` because the `sgs` are of the form: 
```
export interface ISecurityGroupsByAccountSourceData {
    [account: string]: {
        [provider: string]: {
            [region: string]: ISecurityGroupSummary[];
        };
    };
}
```

I'm not how to use `forEach` on this sort of object. 
Would appreciate a suggestion :)

Reply:
I see so you want to check the `key`, as opposed to the values. You could do a truthy check to see if the key exists and then the conditional would return if false preventing any errors. 

```
const sgOptions = sgs.map(sg => 
  if (sg[values.credentials] && sg[values.credentials] ['aws'] && 
  sg[values.credentials] ['aws'][values.region]) {
      const groups = sg[values.credentials] ['aws'][values.region];
      return groups.map({ value: group.id, label: group.name })
  }
 return []
);

return flatten(sgGroups)
```

Reply:
Adding comments for making it easier to understand. Finding  it difficult to break it down as the structure of the object `sgs` is complex. :(